WTRSimUnity/Assets/Scripts/SocketHandler.cs

```diff
@@ -0,0 +1,147 @@
+﻿using System;
+using System.Net;
+using System.Net.Sockets;
+using UnityEngine;
+using UnityEditor;
+using UnityEditor.PackageManager;
+using System.Collections;
+using UnityEngine.Assertions;
+using System.Threading;
+using System.Linq;
+using System.Text;
+
+//Singleton needed on game object
+
+// Credit to the University of Northampton, UK
+// http://www.eng.northampton.ac.uk/~espen/CSY2026/JavaServerCSClient.htm
+// CSY2026 Modern Networks
+
+/**
+ * Class that handles communication with Java
+ */
+[InitializeOnLoad]
+public class SocketHandler: MonoBehaviour
+{
+    int clientPort = 4513;
+    int serverPort = 4512;
+    
+    public string outboundString = "hello Java";
+    public string received;
+
+    Socket sendSocket;
+    UdpClient receiveSocket;
+    IPEndPoint clientEP;
+
+    Boolean dependentMode; //If dependentMode is true, Unity is running dependently to WTRSimlib
+    volatile Boolean kill = false;
+
+    int lastHeartBeat = -1;
+    int heartBeat = 0;
+    Boolean heartStopped = false;
+
+    SocketHandler()
+    {
+       
+        sendSocket = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp);
+        receiveSocket = new UdpClient(clientPort);
+        clientEP = new IPEndPoint(IPAddress.Any, clientPort);
+
+        sendSocket.SendTimeout = 2000;
+        receiveSocket.Client.ReceiveTimeout = 2000;
+
+        try
+        {
+            IPAddress serverIPAddress = IPAddress.Parse("127.0.0.1");
+            IPEndPoint serverEP = new IPEndPoint(serverIPAddress, serverPort);
+            byte[] message = new byte[1];
+            sendSocket.SendTo(message, serverEP);
+
+            lastHeartBeat = heartBeat;
+            receiveSocket.Receive(ref clientEP);
+            heartBeat++;
+
+            Debug.Log("Connected!");
+            dependentMode = true;
+        }
+        catch ( SocketException e )
+        {
+            dependentMode = false;
+        }
+     
+
+        if ( dependentMode )
+        {
+            //Inbound
+            new Thread(() =>
```

Comment:
better to use Update() than to start Thread in constructor

Reply:
Unity Update methods are not real-time. In addition, we don't want to include anything within Unity that may interrupt Unity itself. Update methods of all kinds are called in the main thread of Unity.